File: musiclibrary/src/main/java/com/lzx/musiclibrary/manager/FocusAndLockManager.java
package com.lzx.musiclibrary.manager;

import android.content.Context;
import android.media.AudioManager;
import android.net.wifi.WifiManager;

/**
 * Created by xian on 2018/1/20.
 */

public class FocusAndLockManager {
    private AudioManager mAudioManager;
    private WifiManager.WifiLock mWifiLock;

    public static final float VOLUME_DUCK = 0.2f;
    public static final float VOLUME_NORMAL = 1.0f;

    public static final int AUDIO_NO_FOCUS_NO_DUCK = 0;
    public static final int AUDIO_NO_FOCUS_CAN_DUCK = 1;
    public static final int AUDIO_FOCUSED = 2;

    private int mCurrentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK;

    private AudioFocusChangeListener mListener;

    public FocusAndLockManager(Context context, AudioFocusChangeListener listener) {
        Context applicationContext = context.getApplicationContext();

        this.mAudioManager = (AudioManager) applicationContext.getSystemService(Context.AUDIO_SERVICE);
        this.mWifiLock = ((WifiManager) applicationContext.getSystemService(Context.WIFI_SERVICE))
                .createWifiLock(WifiManager.WIFI_MODE_FULL, "uAmp_lock");
        this.mListener = listener;
    }

    public FocusAndLockManager(Context context) {
        new FocusAndLockManager(context, null);
    }

    public void giveUpAudioFocus() {
        if (mAudioManager.abandonAudioFocus(mOnAudioFocusChangeListener)
                == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            mCurrentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK;
        }
    }

    public void tryToGetAudioFocus() {
        int result =
                mAudioManager.requestAudioFocus(
                        mOnAudioFocusChangeListener,
                        AudioManager.STREAM_MUSIC,
                        AudioManager.AUDIOFOCUS_GAIN);
        if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            mCurrentAudioFocusState = AUDIO_FOCUSED;
        } else {
            mCurrentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK;
        }
    }

    public int getCurrentAudioFocusState() {
        return mCurrentAudioFocusState;
    }

    public void releaseWifiLock() {
        if (mWifiLock.isHeld()) {
            mWifiLock.release();
        }
    }

    public void acquireWifiLock() {
        mWifiLock.acquire();
    }

    private final AudioManager.OnAudioFocusChangeListener mOnAudioFocusChangeListener = new AudioManager.OnAudioFocusChangeListener() {
        @Override
        public void onAudioFocusChange(int focusChange) {
            switch (focusChange) {
                case AudioManager.AUDIOFOCUS_GAIN:
                    mCurrentAudioFocusState = AUDIO_FOCUSED;
                    break;
                case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK:
                    mCurrentAudioFocusState = AUDIO_NO_FOCUS_CAN_DUCK;
                    break;
                case AudioManager.AUDIOFOCUS_LOSS_TRANSIENT:
                    mCurrentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK;

                    if (mListener != null) {
                        mListener.onAudioFocusLossTransient();
                    }
                    break;
                case AudioManager.AUDIOFOCUS_LOSS:
                    mCurrentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK;
                    break;
            }

            if (mListener != null) {
                mListener.onAudioFocusChange();
            }
        }
    };

    public interface AudioFocusChangeListener {
        void onAudioFocusLossTransient();

        void onAudioFocusChange();
    }
}

File: app/src/main/java/com/lzx/musiclib/example/FocusAndLockManager.kt
package com.lzx.musiclib.example

import android.content.Context
import android.media.AudioManager
import android.media.AudioManager.OnAudioFocusChangeListener

import android.net.wifi.WifiManager
import android.net.wifi.WifiManager.WifiLock

/**
 * 焦点管理和WifiLock
 */
class FocusAndLockManager {
    private var mAudioManager: AudioManager? = null
    private var mWifiLock: WifiLock? = null
    var currentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK

    private var mListener: AudioFocusChangeListener? = null

    constructor(context: Context, listener: AudioFocusChangeListener?) {
        val applicationContext: Context = context.applicationContext
        mAudioManager =
            applicationContext.getSystemService(Context.AUDIO_SERVICE) as AudioManager?
        mWifiLock = (applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager)
            .createWifiLock(WifiManager.WIFI_MODE_FULL, "uAmp_lock")
        mListener = listener
    }

    constructor(context: Context) {
        FocusAndLockManager(context, null)
    }

    fun giveUpAudioFocus() {
        if (mAudioManager?.abandonAudioFocus(mOnAudioFocusChangeListener)
            == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            currentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK
        }
    }

    fun tryToGetAudioFocus() {
        val result = mAudioManager!!.requestAudioFocus(
            mOnAudioFocusChangeListener,
            AudioManager.STREAM_MUSIC,
            AudioManager.AUDIOFOCUS_GAIN)
        currentAudioFocusState = if (result == AudioManager.AUDIOFOCUS_REQUEST_GRANTED) {
            AUDIO_FOCUSED
        } else {
            AUDIO_NO_FOCUS_NO_DUCK
        }
    }

    fun releaseWifiLock() {
        if (mWifiLock?.isHeld == true) {
            mWifiLock?.release()
        }
    }

    fun acquireWifiLock() {
        mWifiLock?.acquire()
    }

    private val mOnAudioFocusChangeListener: OnAudioFocusChangeListener =
        OnAudioFocusChangeListener { focusChange ->
            when (focusChange) {
                AudioManager.AUDIOFOCUS_GAIN -> {
                    currentAudioFocusState = AUDIO_FOCUSED
                }
                AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                    currentAudioFocusState = AUDIO_NO_FOCUS_CAN_DUCK
                }
                AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
                    currentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK
                    mListener?.onAudioFocusLossTransient()
                }
                AudioManager.AUDIOFOCUS_LOSS -> {
                    currentAudioFocusState = AUDIO_NO_FOCUS_NO_DUCK
                }
            }
            mListener?.onAudioFocusChange()
        }

    interface AudioFocusChangeListener {
        fun onAudioFocusLossTransient()
        fun onAudioFocusChange()
    }

    companion object {
        const val VOLUME_DUCK = 0.2f
        const val VOLUME_NORMAL = 1.0f
        const val AUDIO_NO_FOCUS_NO_DUCK = 0
        const val AUDIO_NO_FOCUS_CAN_DUCK = 1
        const val AUDIO_FOCUSED = 2
    }
}